aip/general/0211.md

```diff
@@ -0,0 +1,63 @@
+---
+id: 211
+state: reviewing
+created: 2021-02-24
+placement:
+  category: design-patterns
+  order: 115
+---
+
+# Authorization checks
+
+The majority of operations, whether reads or writes, require authorization:
+permission to do the thing the user is asking to do. Additionally, it is
+important to be careful how much information is provided to _unauthorized_
+users, since leaking information can be a security concern.
+
+## Guidance
+
+Services **must** check authorization before validating any request, to ensure
+both a secure API surface and a consistent user experience. An operation
+**may** require multiple permissions or preconditions in order to grant
+authorization.
+
+If a request can not pass the authorization check for any reason, the service
+**must** error with `PERMISSION_DENIED`, and the corresponding error message
+**should** look like: "Permission `{p}` denied on resource `{r}` (or it might
+not exist)." This avoids leaking resource existence.
```

Comment:
Consider the following scenario

* User1 has access to Get `books/*/category/scifi` but not any other category
* `books/*/category/secret-category` resource exists
* `books/*/category/not-existant-category` does not exist

If I understood the guidance correctly:

1. User1 tries to Get `books/*/category/secret-category` receives PERMISSION_DENIED
2. User1 tries to Get `books/*/category/not-existant-category` receives PERMISSION_DENIED

Is this correct?

Reply:
Yes, and the error message says "Permission denied to `not-existant-category` (or it may not exist)."